### ISSUE TITLE
Document and Research - How to create an Act Once operators in the Toolshelf

### ISSUE THREAD

**Issue by Draise14** [Bforartists/Bforartists#2162]
# Objective
![image](https://user-images.githubusercontent.com/35357409/104210046-68ce0f00-5400-11eb-8c9f-3d9dd999691e.png)

Currently we have a lot of act once operators in header menus and sub menus. They may also be in a right click menu.. but they are not exposed in a top level mouse over and click anywhere. 

We could ultimately expose these operators as a customization in the toolbar per Bforartists 1 method, but this requires re-writing many hundreds of classes. 

Research on that done here with a rotate tool from the UV Editor:
#2154

-------

The alternative is adding them to the toolshelf with panels, also as an optional idea, with a panel based system like from here:
#1395

This is another research task. 

------

But this task is to find out... first and foremost, how do we make an act once operator in the toolshelf?

Here are some clues as to how this might be done:
#2138

# Task
Do Research and Document how: 
_Find a way to create act once operators where you can click it and it will act once, then revert to former toolshelf mode tool._ 
- Document which files need changes
- Document how to do the code
- Create a mockup and a showcase tool with an existing header based operator, icon and then placement in toolshelf
- Create a quick tutorial on how to do it
- Evaluate possibilities

## Method
I've asked @iyadahmed to do some research into this, document it and find away to save some time figuring it out to give us some options.

**Comment by ReinerBforartists** [comment]
Don't we have a task for this already? This is part of the research how to implement back the tabs into the tool shelf, especially for the old operators that shall not go into the toolbar with the 24px icons. 

We had experiments to append normal panels in the toolshelf branch already. This works. The structure is still the same. But it looks odd to have a new style panel and an old style panel underneath. Thus the idea to add the tabs back.

https://github.com/Bforartists/Bforartists/issues/1402

**Comment by Draise14** [comment]
Ah, this task is tabs agnostic, just to make it smaller and easier to manage.. first step, just learn how to make act once operators in the toolshelf with the new icon system for UI and design consistency. 

We could do this using the old icons and panels, which are a lot smaller. But not quite refined enough.

**Comment by ReinerBforartists** [comment]
We have this done already. And decided that we need tabs here. See answer above.

**Comment by Draise14** [comment]
But we haven't figured this out, this task is for this: 
![image](https://user-images.githubusercontent.com/35357409/104214027-f7448f80-5404-11eb-8fda-59c311c7eb52.png)

**Comment by ReinerBforartists** [comment]
Because this doesn't work. New tool system has big buttons and completely different code.

**Comment by Draise14** [comment]
I guess we can postpone this to try tackle tabs first so we can justify panels and more buttons, I get it - but we still wanna get the code to work with act once in the one and same toolshelf. Thus this task.

**Comment by Draise14** [comment]
Postponed for this task as priority:
#1402

**Comment by ReinerBforartists** [comment]
Act once is not the new toolshelf. The new toolshelf is permanent tools. We need to keep them separate. When we start to mix the concepts then we get a big problem. Isn't technically possible anyways when i remember the code right. 

We have for example this task here, which requires permanent toolshelf buttons: https://github.com/Bforartists/Bforartists/issues/1339

**Comment by ReinerBforartists** [comment]
Looking at the image again, we do have some of these transform tools already in the toolshelf. But not as one trick tools. They are integrated as permanent tools.

But i would highly suggest to add this functionality at a one by one base, tool by tool. This is a complicated can of worms, you need to take modes and object types into account. And as permanent tools, not one trick. As told, either or.

We would also need new icons for the toolshelf then. We have still an open task to find out if we could bypass the nasty Blender way of icon creation. I am not happy to model the icons to fit the pipeline when i have the png files already finished.

![transform](https://user-images.githubusercontent.com/13919702/104215171-b49caa80-5438-11eb-9e2e-7aaac31e7c19.jpg)

**Comment by Draise14** [comment]
I agree. Case per case. This task is to learn how with a mockup of ONE operator.. then it closes. 

Then we make new ones on a case per case basis.

**Comment by Draise14** [comment]
Just to clarify, I'm agreeing with you here. This task is just to try find a way to get act once to the new toolshelf icons and buttons. It is technically possible to a degree according to some experiments on Blender stack exchange - so there must be a way. This is just to see/research if this is possible, officially. I'll have @iyadahmed take a look at it. Once a single mockup is done, this task closes. If it is confirmed that this can not be done, this will close too. Just wanna give him a shot! 

Concerning UX, if this is a success, defining what is modal and act-once would be set by the Icon and tool group colours, imo. Or an icon colour or extra symbol. But at the end of the day, they both are tools.. for a shelf. And that would be another set of tasks/subtasks per operator. Not all of them would go to the toolshelf, ei, Bforartists 1 priority tools vs. secondary tools stored in header. 

As you mentioned.. we can't have two kinds of tool buttons in the same toolshelf, though we ultimately want all/most tools in the toolshelf per Bforartists 1 workflows. This is to try find a way to do that, mix act once with modal in the same system. The UX can be sorted in how we organize and show the iconography to make sure the UX is clean later.

**Comment by ReinerBforartists** [comment]
Of course it is possible. You just need to look at the commits that adds new tools into the toolshelf. There is no better explanation. I don't need extra research for this part. Also not for the old operators. We have also an example that we can add a panel with the old operators below the toolshelf tools. The culprit in the toolshelf is imho to add back the tabs so that we can put the panels with the old operators into the toolshelf too, each in its own tab.

It's just we haven't arrived at the toolshelf yet. But you can of course ask Iyad if he wants to do research here.

I am not happy to add modal operators in the permanent way. We will run into trouble here. We have already a good way to separate them by simply using the old style panels for the modal operators. And the new icon buttons for the permanent toolset.

**Comment by ReinerBforartists** [comment]
This task is tied to if we ever manage to reimplement the tabs in the tool shelf. 

Toolshelf - readd tabs #1402

**Comment by ReinerBforartists** [comment]
I have to close this task. There is no clear task. But more a bunch of tasks mixed with questsion. And this is not to solve.

It is not forgotten. But as told, depends at if we ever manage to reimplement the tabs. Corresponding tasks will appear automatically then. One step after another.

**Comment by Draise14** [comment]
No problem.
